marlowe-protocols/src/Network/Protocol/Connection.hs

```diff
@@ -0,0 +1,207 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+
+module Network.Protocol.Connection
+  where
+
+import Control.Applicative ((<|>))
+import Control.Concurrent.STM (STM, TQueue, atomically, newTQueue, readTQueue, writeTQueue)
+import Control.Exception (SomeException)
+import Control.Monad.Base (MonadBase, liftBase)
+import Control.Monad.Cleanup (MonadCleanup)
+import Control.Monad.Trans.Control (MonadBaseControl)
+import Data.ByteString.Lazy (ByteString)
+import Data.ByteString.Lazy.Base16 (encodeBase16)
+import qualified Data.Text.Lazy as T
+import Data.Void (Void)
+import Network.Channel
+  (Channel(..), ChannelSelector, STMChannel(..), channelPair, getChannelSelectorConfig, hoistChannel, logChannel)
+import Network.Protocol.Codec (BinaryMessage)
+import Network.Protocol.Peer (PeerSelector, getPeerSelectorConfig, logPeer)
+import Network.TypedProtocol
+import Observe.Event (EventBackend, failEvent, finalize, newEvent, subEventBackend)
+import Observe.Event.Backend (narrowEventBackend)
+import Observe.Event.Component (GetSelectorConfig, SelectorConfig(..), absurdFieldConfig, prependKey)
+import Observe.Event.Network.Protocol (MessageToJSON)
+
+type ToPeer peer protocol pr st m = forall a. peer m a -> Peer protocol pr st m a
+
+newtype Connector ps pr peer m = Connector
+  { connectPeer :: forall a. peer m a -> m (Connection ps pr m a)
+  }
+
+type ClientConnector ps = Connector ps 'AsClient
+type ServerConnector ps = Connector ps 'AsServer
+
+data SomeConnector pr peer m =
+  forall ps. BinaryMessage ps => SomeConnector (Connector ps pr peer m)
+
+type SomeClientConnector = SomeConnector 'AsClient
+type SomeServerConnector = SomeConnector 'AsServer
+
+newtype ConnectionSource ps server m = ConnectionSource
+  { acceptConnector :: STM (Connector ps 'AsServer server m)
+  }
+
+mergeConnectionSource :: ConnectionSource ps server m -> ConnectionSource ps server m -> ConnectionSource ps server m
+mergeConnectionSource (ConnectionSource source1) (ConnectionSource source2) = ConnectionSource $ source1 <|> source2
+
+data SomeConnectionSource server m =
+  forall ps. BinaryMessage ps => SomeConnectionSource (ConnectionSource ps server m)
+
+data ConnectorSelector ps f where
+  Session :: ConnectorSelector ps Void
+  ConnectionSelector :: ConnectionSelector ps f -> ConnectorSelector ps f
+
+getConnectorSelectorConfig
+  :: MessageToJSON ps
+  => Bool
+  -> Bool
```

Comment:
The `Bool`s are a bit opaque when these functions are called. Maybe it is worthwhile to add a comment in the two places where these are called with `False`?